PIPS/pip-1.md

```diff
@@ -0,0 +1,147 @@
+
+This PIP proposes an alternative implementation which does not require an operator.
+The tradeoff of this implementation is that an asset can be transferred between at most two chains.
+
+:white_check_mark: Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ethereum LockProxy => Unlock token A
+
+:x: Not Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ontology LockProxy => Unlock token A
+3. Lock token A => Ontology LockProxy --> Ethereum LockProxy => This will fail
+
+## Specification
+To support non-native tokens within the LockProxy, a token contract should be deployed to represent the non-native token.
+
+The token's constructor should call `delegateAsset` on the LockProxy that it is delegated to, for example:
+
+```
+constructor (address lockProxyContractAddress, uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash) public ERC20Detailed("ONT Token", "ONTX", 0) {
+    uint256 totalSupply = 1000000000;
+    _mint(lockProxyContractAddress, totalSupply);
+    lockProxy.delegateAsset(nativeChainId, nativeLockProxy, nativeAssetHash, totalSupply);
+}
+```
+
+In the LockProxy, `delegateAsset` can be implemented as:
+```
+struct DelegatedAsset {
+    uint64 nativeChainId;
+    bytes nativeLockProxy;
+    bytes nativeAssetHash;
+}
+
+mapping(address => DelegatedAsset) delegatedAssets;
+mapping(bytes32 => uint256) balances;
+
+function delegateAsset(uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash, uint256 totalSupply) public {
+    address assetHash = _msgSender();
+
+    require(nativeChainId != 0);
+    require(delegatedAssets[assetHash].nativeChainId == 0);
+
+    // `hash` is any supported hashing function of the respective blockchain
+    bytes32 key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(balances[key] == 0);
+    balances[key] = totalSupply;
+
+    delegatedAssets[_msgSender()] = DelegatedAsset(nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    // call registerAsset here for convenience
+    // registerAsset is explained in a later section
+    registerAsset(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+This mapping will be used in the `unlock` function to ensure that delegated tokens will be unlocked only if the specified token is locked.
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed, to replace them we can add a `registerAsset` function:
+
+```
+mapping(bytes32 => bool) registry;
+
+function registerAsset(address fromAssetHash, uint64 toChainId, bytes memory targetProxyHash, bytes memory toAssetHash) {
```

Comment:
I think, by design, this method should only be invoked by the delegateAsset() method, so maybe need to be declared as internal function.

Reply:
> actually I think we can remove `registerAsset` completely since `delegateAsset` will already update a `delegatedAssets` mapping that performs the same check, I'll remove it

I will just remove this function for now as I think it is not needed